File: plugins/collector/event/apps/database/etcd.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Etcd struct {
	version string
	reg     *regexp.Regexp
}

func (Etcd) Name() string { return "etcd" }

func (Etcd) Type() string { return "database" }

func (e Etcd) Version() string { return e.version }

func (Etcd) Match(p *process.Process) bool { return p.Name == "etcd" }

func (e *Etcd) Run(p *process.Process) (mapping map[string]string, err error) {
	if e.reg == nil {
		e.reg = regexp.MustCompile(`etcd Version: (\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "--version")
	if err != nil {
		return nil, err
	}
	str := e.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	e.version = strings.TrimPrefix(str, "etcd Version: ")
	return
}

func init() { apps.Regist(&Etcd{}) }

File: plugins/collector/src/event/apps/database/etcd.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Etcd {
    version: String,
}

impl Etcd {
    pub fn new() -> Self { Etcd { version: String::new() } }
}

impl IApp for Etcd {
    fn name(&self)     -> &'static str { "etcd" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "etcd" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "etcd Version: 3.5.5"
        for line in out.lines() {
            if let Some(rest) = line.strip_prefix("etcd Version:") {
                self.version = rest.trim().to_owned();
                return Ok(HashMap::new());
            }
        }
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
